src/spdx_tools/spdx/model/document.py

```diff
@@ -81,3 +81,6 @@ def __init__(
         relationships = [] if relationships is None else relationships
         extracted_licensing_info = [] if extracted_licensing_info is None else extracted_licensing_info
         check_types_and_set_values(self, locals())
+
+    def __hash__(self):
+        return id(self)
```

Comment:
documents are not immutable and therefore hashing might run into inconsistencies. Most use-cases would probably expect a content dependent hash.

Reply:
Ah that's a shame.

I wouldn't expect for the document to get mutated during validation. Would it make sense to have some frozen/immutable variants of document and relationship that could be passed around the validators?

Or do you have any ideas that would be a better approach to take?

Reply:
Agree that a hash should only be implemented on immutable objects _when it uses mutable properties of the object_ - most "useful" hashes do. This hash is not based on a mutable property which would make it safe and consistent although not very useful (The default implementation of __hash__ is just id(self) >> 4). Also note that __eq__ should always be defined when __hash__ is [[see here](https://docs.python.org/3.5/reference/datamodel.html#object.__hash__)]. Why define __hash__ here?

Reply:
I've used `functools.lru_cache` in the past to speed up slow loops; it watches the parameters and replaces a function call with a table lookup when that makes sense.  The nice thing is that it can catch mutations and know to call the function again without calling it every single time.  Not sure if that would help in this case; I'll try to take a look this weekend.